client/object_search.go

```diff
@@ -243,75 +233,37 @@ func (c *Client) ObjectSearchInit(ctx context.Context, prm PrmObjectSearch) (*Ob
 		panic(panicMsgMissingContainer)
 	}
 
-	// form request body
-	var (
-		body  v2object.SearchRequestBody
-		cidV2 v2refs.ContainerID
-	)
-
+	var cidV2 v2refs.ContainerID
 	prm.cnrID.WriteToV2(&cidV2)
 
+	var body v2object.SearchRequestBody
 	body.SetVersion(1)
 	body.SetContainerID(&cidV2)
 	body.SetFilters(prm.filters.ToV2())
 
-	// form meta header
-	var meta v2session.RequestMetaHeader
-
-	if prm.local {
-		meta.SetTTL(1)
-	}
+	// init reader
+	var req v2object.SearchRequest
+	req.SetBody(&body)
+	c.prepareRequest(&req, &prm.meta)
 
-	if prm.bearerSet {
-		var v2token acl.BearerToken
-		prm.bearer.WriteToV2(&v2token)
-		meta.SetBearerToken(&v2token)
+	key := prm.key
+	if key == nil {
+		key = &c.prm.key
 	}
 
-	if prm.sessionSet {
-		var tokv2 v2session.Token
-		prm.session.WriteToV2(&tokv2)
-
-		meta.SetSessionToken(&tokv2)
+	err := signature.SignServiceMessage(key, &req)
+	if err != nil {
+		return nil, fmt.Errorf("sign request: %w", err)
 	}
 
-	prm.prmCommonMeta.writeToMetaHeader(&meta)
-
-	// form request
-	var req v2object.SearchRequest
-
-	req.SetBody(&body)
-	req.SetMetaHeader(&meta)
-
-	// init reader
-	var (
-		r      ObjectListReader
-		resp   v2object.SearchResponse
-		stream *rpcapi.SearchResponseReader
-	)
-
+	var r ObjectListReader
 	ctx, r.cancelCtxStream = context.WithCancel(ctx)
 
-	resp.SetBody(&r.bodyResp)
-
-	// init call context
-	c.initCallContext(&r.ctxCall)
-	r.ctxCall.req = &req
-	r.ctxCall.statusRes = new(ResObjectSearch)
-	r.ctxCall.resp = &resp
-	r.ctxCall.wReq = func() error {
-		var err error
-
-		stream, err = rpcapi.SearchObjects(&c.c, &req, client.WithContext(ctx))
-		if err != nil {
-			return fmt.Errorf("open stream: %w", err)
-		}
-
-		return nil
-	}
-	r.ctxCall.rResp = func() error {
-		return stream.Read(&resp)
+	r.stream, err = rpcapi.SearchObjects(&c.c, &req, client.WithContext(ctx))
```

Comment:
just wanna highlight that change one more time: we now open stream in the init call, not in a reader. however it look more like a fix cause all our stream RPC has `The call only opens the transmission channel, explicit recording is done using the *.`

could lead to some unexpected changes in the clients (mb SN)

Reply:
Yes, I have researched a bit -- pool and neofs-cli use `Init` immediately followed by `WriteHeader`, so everything should be ok.

Reply:
Actually, `opens the transmission channel` message shouldn't even be changed -- this is what we do here, open the stream for further processing.

Reply:
> shouldn't even be changed

yea, that is why i called it more like a fix not a refactor in that case